### DISCUSSION TITLE
过期：缩略图脚本中 自动删除的多实例兼容问题

### DISCUSSION THREAD

**Discussion by zhongfly** [hooke007/mpv_PlayKit#48] [回收]
问题最初在论坛中有人提出 https://bbs.acgrip.com/forum.php?mod=redirect&goto=findpost&ptid=5843&pid=77986&fromuid=12177

> 同时打开两个视频窗口播放视频，且都建立了缩略图，当关闭其中一个窗口时，会导致另一个窗口的缩略图变没。

主要原因是缩略图脚本将生成的缩略图输出到固定的文件夹，假如设置关闭时清除的话，脚本会在关闭时直接删除整个文件夹。这使得还未关闭窗口的缩略图也被删除。

我也和脚本作者询问过这个情况，https://github.com/deus0ww/mpv-conf/issues/15
我提出了2种简单的方法：
1. 每个窗口都使用单独的文件夹里存放缩略图，不过作者认为会影响到“不设置自动删除，重复利用已有缩略图”的情况。
但我觉得依然可行，根据是否自动删除来判断要不要建立独立文件夹（自动删除时，每个窗口根据pid建立独立文件夹；不自动删除时与现在一致）
2. 每个窗口用一个列表记录自己建立的缩略图，关闭时只删除该窗口建立的缩略图
但是作者不愿意这么做，但我自己修改了，没啥大问题。
https://github.com/zhongfly/mpv-conf/commit/5c6fa181c55a9190559669742c1dfe6f4c7651c2#diff-460fb2c70b9e5208658d3f7f1cea5c096cb9100a98316c4539d41aa019fb0edd
不过仍存在一个问题，就是2个窗口打开同一个视频，关闭一个窗口、自动删除时自然还是会删除该视频的缩略图，另一个窗口就又出现缩略图丢失的问题了。
是否能实现类似文件锁的东西，只要还有窗口正在使用这个文件夹（正在播放对应的视频），就不会被删除

**Comment by zhongfly** [comment]
关于第2个方法 想到了改进，在视频对应的缩略图文件夹里加一个lock文件夹，打开一次该视频(建立缩略图)，就往里面增加一个名字为当前进程pid的文件，删除时先在lock文件夹里删除当前pid名的文件，然后判断lock文件夹是否为空(utils.readdir)…

**Reply by zhongfly** [reply]
方法2可能的隐患：由于是依次删除多个文件夹，可能会出现关闭mpv的过程被卡住，亦或者是在mpv.net这类gui使用中没等全删除完就被强行中断。
在mpv.net上测试发现，删除性能堪忧，退出时只来得及删除9个视频的缩略图

**Comment by hooke007** [comment]
同一个视频不同进程无法复用同样的缩略图缓存造成的性能浪费太多了。
我对这个脚本体感最糟糕的还是性能（尽管已经比AMM原始版本强太多了）。其他倒是次要的。看电影的时候真的风扇呼呼转的又久又要命，还卡。。。

**Reply by zhongfly** [reply]
我提出的第2个方法就可以复用啊…性能方面我倒没觉得有任何问题。

**Reply by hooke007** [reply]
忘了写针对序号1

**Comment by zhongfly** [comment]
按第二种方法，加了文件锁，如果其余窗口还在使用某缩略图时，就会保留。
https://github.com/zhongfly/mpv-conf/blob/main/scripts/Thumbnailer.lua

问题就是在mpv.net中，设置关闭时清理缩略图，关闭窗口时只来得及删除9个视频的缩略图（根据我的电脑进行测试），使得不能够清理完全……除非多整几个worker专门负责清理缩略图

**Reply by zhongfly** [reply]
瓶颈在于删除文件夹的操作……

**Comment by dyphire** [comment]
题外话：多窗口打开同视频的场景是什么？

**Reply by zhongfly** [reply]
（不小心重复打开了视频

**Comment by zhongfly** [comment]
想出了方法3：
自动删除分为2种：
1. 文件关闭时自动删除，此时在每个视频的缩略图文件夹里放置文件锁（名称为pid），关一个文件，删除本进程的锁，检查是否无锁，删除一个缩略图文件夹
2. 退出时自动删除，此时在缩略图的根文件夹处放置文件锁（名称为pid），关闭播放器时，删除本进程的锁，检查这个总锁，删除整个缓存文件夹

https://github.com/zhongfly/mpv-conf/blob/main/scripts/Thumbnailer.lua 
edit： 修正设置“文件关闭时自动删除”时，退出播放器的误操作

其实还有一种情况就是播放中途通过快捷键切换“自动删除”选项导致的问题，不过这样的情形也太少见了。不过还是略微修改了一下

**Reply by dyphire** [reply]
这个方法不错，检测最后一个进程退出时删除整个缓存文件夹，原有操作逻辑的增强

**Reply by zhongfly** [reply]
> 这个方法不错，检测最后一个进程退出时删除整个缓存文件夹，原有操作逻辑的增强

修了一个bug
设置“文件关闭时自动删除”：当退出播放器时，会先触发事件“文件关闭”，于是调用逻辑1删除**子**文件夹，之后触发事件2“退出”，调用逻辑2，由于此时根目录没有文件锁，则必定会删除整个缓存文件夹，不管有没有其他进程使用

**Reply by zhongfly** [reply]
@hooke007 你觉得如何，有没有必要添加到仓库里的脚本？

**Reply by hooke007** [reply]
你可以直接给我推pr（
